File: Problems/20/main.js
/**
 * @param {string} s
 * @return {boolean}
 */
const isValid = (s) => {
  const stack = [];
  for (let char = 0; char < s.length; char++) {
    if (s[char] === "(" || s[char] === "{" || s[char] === "[") {
      stack.push(s[char]);
    } else if (s[char] === ")") {
      if (stack[stack.length - 1] !== "(") {
        return false;
      } else {
        stack.pop();
      }
    } else if (s[char] === "]") {
      if (stack[stack.length - 1] !== "[") {
        return false;
      } else {
        stack.pop();
      }
    } else if (s[char] === "}") {
      if (stack[stack.length - 1] !== "{") {
        return false;
      } else {
        stack.pop();
      }
    }
  }
  if (stack.length === 0) {
    return true;
  }
  return false;
};

// TEST SUITE

// TEST 1

function test1() {
  const actual = isValid("()");
  const expected = true;
  if (actual === expected) {
    return "$$$$$";
  }
  return "X";
}

console.log(test1());

// TEST 2

function test2() {
  const actual = isValid("()[]{}");
  const expected = true;
  if (actual === expected) {
    return "$$$$$";
  }
  return "X";
}

console.log(test2());

// TEST 3

function test3() {
  const actual = isValid("(]");
  const expected = false;
  if (actual === expected) {
    return "$$$$$";
  }
  return "X";
}

console.log(test3());

// TEST 4

function test4() {
  const actual = isValid("([)]");
  const expected = false;
  if (actual === expected) {
    return "$$$$$";
  }
  return "X";
}

console.log(test4());

// TEST 5

function test5() {
  const actual = isValid("{[]}");
  const expected = true;
  if (actual === expected) {
    return "$$$$$";
  }
  return "X";
}

console.log(test5());

// TEST 6

function test6() {
  const actual = isValid("[");
  const expected = false;
  if (actual === expected) {
    return "$$$$$";
  }
  return "X";
}

console.log(test6());

// TEST 7

function test7() {
  const actual = isValid("(){}}{");
  const expected = false;
  if (actual === expected) {
    return "$$$$$";
  }
  return "X";
}

console.log(test7());

File: Problems/20/main.py
def isValid(s):
    stack = []
    if s[0] == ')' or s[0] == ']' or s[0] == '}':
        return False
    count = 0
    for char in s:
        count += 1
        if char == '(' or char == '[' or char == '{':
            stack.append(char)
        elif char == ')':
            if len(stack) == 0:
                return False
            if stack[len(stack) - 1] != '(':
                return False
            else:
                stack.pop()
        elif char == ']':
            if len(stack) == 0:
                return False
            if stack[len(stack) - 1] != '[':
                return False
            else:
                stack.pop()
        elif char == '}':
            if len(stack) == 0:
                return False
            if stack[len(stack) - 1] != '{':
                return False
            else:
                stack.pop()
    if len(stack) == 0:
        return True
    return False

# TEST SUITE

# TEST 1

def test1():
    actual = isValid("()")
    expected = True
    if actual == expected:
        return "$$$$$"
    return "X"

print(test1())

# TEST 2

def test2():
    actual = isValid("()[]{}")
    expected = True
    if actual == expected:
        return "$$$$$"
    return "X"

print(test2())

# TEST 3

def test3():
    actual = isValid("(]")
    expected = False
    if actual == expected:
        return "$$$$$"
    return "X"

print(test3())

# TEST 4

def test4():
    actual = isValid("([)]")
    expected = False
    if actual == expected:
        return "$$$$$"
    return "X"

print(test4())

# TEST 5

def test5():
    actual = isValid("{[]}")
    expected = True
    if actual == expected:
        return "$$$$$"
    return "X"

print(test5())

# TEST 6

def test6():
    actual = isValid("[")
    expected = False
    if actual == expected:
        return "$$$$$"
    return "X"

print(test6())

# TEST 7

def test7():
    actual = isValid("]")
    expected = False
    if actual == expected:
        return "$$$$$"
    return "X"

print(test7())

# TEST 8

def test8():
    actual = isValid("(){" + "}" + "}{") # Divided to avoid python string formatting behavior
    expected = False
    if actual == expected:
        return "$$$$$"
    return "X"

print(test8())
